sympy/combinatorics/fp_groups.py

```diff
@@ -297,33 +297,66 @@ def __init__(self, G, gens):
         super(FpSubgroup,self).__init__()
         self.parent = G
         self.generators = list(set([g for g in gens if g != G.identity]))
+        self._min_words = None #for use in __contains__
         self.C = None
 
     def __contains__(self, g):
-        gens = [s for s in g.contains_generators()]
-        if any([s not in self.generators for s in gens]):
+        if self._min_words is None:
+            gens = self.generators[:]
+            gens.extend([e**-1 for e in gens])
+            for i, w1 in enumerate(gens):
+                for w2 in gens[i+1:]:
+                    if w2**-1 == w1:
+                        continue
+                    if w2[len(w2)-1]**-1 == w1[0] and w2*w1 not in gens:
+                        gens.append(w2*w1)
```

Comment:
I didn't use a set because looping over it consistently would be hard. I'm not quite sure what you are suggesting (the `remove one of its elements w1 and append it to min_words before continuing` part) but I understand what the potential problem is. I'll try to fix it.